Create contents list component

fofr opened the pull request:
Marked as do not merge while I double check cross browser. Ready for review.

Notes to reviewers:
* Preview: http://government-frontend-pr-420.herokuapp.com/component-guide/contents-list
* Looking for guidance on accessibility acceptance criteria to use
* Please check my BEM

Changes:
* Switch all formats to use contents list component
* Consolidates specialist documents with same pattern as other formats
* Remove analytics tracking on contents lists, currently unused
* Remove unused templates
* Incorporates @andysellick’s contents list suggestions in this PR: https://github.com/alphagov/government-frontend/pull/394

![contents list component - government frontend component guide 20170801](https://user-images.githubusercontent.com/319055/28833673-f1b00cce-76d7-11e7-88e2-98ab16bae7cf.png)

vanitabarrett commented on app/views/components/docs/contents-list.yml:
```diff
@@ -0,0 +1,100 @@
+  - accept focus
+  - be focusable with a keyboard
+  - be usable with a keyboard
+  - indicate when it has focus
```
How about must inform the user how many items are in the list?

andysellick replied:
Should we also have a criteria about informing the user when the list contains a sub-list? I don't think it's something we'll need to add to the markup, Voiceover already announces that there's another list within a list.

fofr replied:
Added

vanitabarrett commented on app/views/components/docs/contents-list.yml:
```diff
@@ -0,0 +1,100 @@
+  `format_numbers` option will pull out numbers in the link text to render them as though they were the list style type. Applies to
+  numbers at the start of text, with or without a decimal. See the [format complex numbers fixture](/component-guide/contents-list/formats_complex_numbers) for details.
+accessibility_criteria: |
+  The contents links must:
```
Should this be 'contents list'? Or are you writing criteria for links within the list? If so, maybe 'contents list links' might make that clearer.

fofr replied:
Updated

andysellick commented on app/views/components/docs/contents-list.yml:
```diff
@@ -0,0 +1,100 @@
+  formats_complex_numbers:
+    format_numbers: true
+    contents:
+      - href: "#"
```
Is there a reason that all the other fixtures have a full href value like #first-thing but the ones in this one are just #?

fofr replied:
No reason.

andysellick commented on app/views/components/docs/contents-list.yml:
```diff
@@ -0,0 +1,100 @@
+      - href: "#"
+        text: 2017 four digit numbers ignored
+      - href: "#"
+        text: "2001: A space odyssey"
```
😂

andysellick commented on test/components/contents_list_component_test.rb:
```diff
@@ -0,0 +1,65 @@
+    assert_select "#{nested_link_selector} .app-c-contents-list__number", count: 0
+    assert_select "#{nested_link_selector} .app-c-contents-list__numbered-text", count: 0
+  end
+end
```
Do we need an explicit test for calling the component with format_numbers true when there are no numbers to format?

fofr replied:
I don't think so.

andysellick commented on app/views/components/docs/contents-list.yml:
```diff
@@ -0,0 +1,100 @@
+        - href: "#second-thing"
+          text: No numbers here
+        - href: "#third-thing"
+          text: None here either
```
Ideally in the component guide it would be good to show the component in a right to left context. Perhaps rather than having a specific fixture option for this we could have the gem do it automatically, so each example gets repeated in RTL? Although that might throw up problems if there are unique IDs in those examples, maybe link to it separately.

Not really a comment on this PR, admittedly, although this component would definitely benefit from such a feature.

vanitabarrett replied:
I think there should already be a story for this in the backlog/planning board? It came up when doing the translation component too.

fofr commented on app/views/components/_contents-list.html.erb:
```diff
@@ -0,0 +1,42 @@
+      <% contents.each do |contents_item| %>
+        <li
+          class="app-c-contents-list__list-item app-c-contents-list__list-item--<%= parent_list_item_modifier %>"
+        >
```
I wanted to avoid overriding default styles in the modified class.

fofr commented on app/views/components/_contents-list.html.erb:
```diff
@@ -0,0 +1,42 @@
+
+%>
+<% if contents.any? %>
+  <nav role="navigation" class="app-c-contents-list">
```
Some guidance here:
https://accessibility.blog.gov.uk/2016/05/27/using-navigation-landmarks/

> We know that some people use the GOV.UK website with older browsers and screen readers, so we use both the HTML5 elements and ARIA roles to make sure the landmarks are available to as many people as possible. A good example is Internet Explorer 11, which does not have accessibility support for the HTML5 elements used to provide landmarks, but which does support the equivalent ARIA roles.

andysellick commented:
IE8 is doing a weird thing with the focus state for dashed items in the contents list. Only happens for list items with the dash. See screenshot. Not occurring in IE9. Dashed items on live site in IE8 currently appear as regular dotted list items.

<img width="230" alt="screen shot 2017-08-07 at 11 44 49" src="https://user-images.githubusercontent.com/861310/29023514-e3b4d18a-7b65-11e7-8a57-e8fe9ed3b88b.png">

andysellick commented:
Firefox doesn't underline some of the elements correctly in the focus state (not a big problem, just being thorough). Latest version.

Chrome (correct):

<img width="905" alt="screen shot 2017-08-07 at 12 19 03" src="https://user-images.githubusercontent.com/861310/29024589-aad74604-7b6a-11e7-9b3d-5f4571516330.png">

Firefox (um):

<img width="899" alt="screen shot 2017-08-07 at 12 19 12" src="https://user-images.githubusercontent.com/861310/29024599-b29d6af8-7b6a-11e7-8a09-3e440078713a.png">

andysellick commented:
Firefox (latest) does some more odd things on numbered items (only) when focussed.

<img width="308" alt="screen shot 2017-08-07 at 12 20 44" src="https://user-images.githubusercontent.com/861310/29024645-ed853470-7b6a-11e7-9c22-db5eaf8c3aec.png">

<img width="346" alt="screen shot 2017-08-07 at 12 20 56" src="https://user-images.githubusercontent.com/861310/29024651-f2bfa182-7b6a-11e7-82d2-36185c2fa56f.png">

fofr commented:
@andysellick Pushed a fix for IE8 focus styles in https://github.com/alphagov/government-frontend/pull/420/commits/d4c2044da33af2c19cde7ad813b6a9f22aab2d7a if you want to re-test.

fofr commented on app/assets/stylesheets/components/_contents-list.scss:
```diff
@@ -0,0 +1,67 @@
+  padding-right: $contents-spacing;
+
+  &:before {
+    content: "— ";
```
Tried doing this quickly. It's a little fiddly. We want the em dash to be in the top left and not have text wrapping beneath it, which means wrapping it in a span. Then we need some logic to include/exclude that span based on whether dashes are being used. It felt wrong moving inheritance out of the stylesheet.

fofr commented on app/views/components/_contents-list.html.erb:
```diff
@@ -0,0 +1,42 @@
+                              else
+                                'dashed'
+                              end
+
```
Fixed

fofr commented:
Requires some fixes for print styles:

![screen shot 2017-08-07 at 13 57 28](https://user-images.githubusercontent.com/319055/29027388-6753c9da-7b78-11e7-8d5c-bf52c6616ac2.png)

fofr commented:
Fixed print styles in 3a5457877fdd4b383c439f0a67376c3795562cb4 and added to component guide in https://github.com/alphagov/govuk_publishing_components/pull/19

Print styles:
[Contents list component - Government Frontend Component Guide.pdf](https://github.com/alphagov/government-frontend/files/1205043/Contents.list.component.-.Government.Frontend.Component.Guide.pdf)

fofr commented:
Print styles can now be reviewed at:
http://government-frontend-pr-420.herokuapp.com/component-guide/contents-list

nickcolley commented:
Checked the print styles, looks good 👍
